### PR TITLE
Caddy archive from ansible controller

### LINKED ISSUE #39: Caddy archive passed from controller rather than downloaded
At the moment caddy-ansible downloads the Caddy archive from the Web on the remote system (as opposed to the Ansible controller). In order to make the deployment more predictable, I am considering downloading Caddy on the Ansible controller and then copying it over.

Before I put effort on this, would you welcome such a contribution?

### PR DESCRIPTION
**Pull request by lucafavatella**
Closes #39
Follows https://github.com/antoiner77/caddy-ansible/pull/41#issuecomment-290674205

### CODE REVIEW

**Comment by lbischof** [comment]
Thanks, I will need to look at this thoroughly when I have more time. It might be worth writing a few tests for Travis or Vagrant because the logic is getting quite complex.

**Comment by lucafavatella** [comment]
> Thanks, I will need to look at this thoroughly when I have more time.

Thanks, no hurry.

> It might be worth writing a few tests for Travis or Vagrant because the logic is getting quite complex.

I plan not to write tests - no bandwidth at the moment.

Please let me know if you prefer a separate PR for commit https://github.com/antoiner77/caddy-ansible/pull/42/commits/42364a9faf8152260413d3d2a6aea49fdd9bcc20

**Comment by lbischof** [comment]
Sorry, I still didn't get to reviewing and testing your PRs. I have exams next week, but after that I should have time.

File: tasks/main.yml [hunk 1]
```diff
@@ -16,15 +20,21 @@
 
 - name: Download Caddy
   get_url: url=https://caddyserver.com/download/build?os=linux&arch=amd64&features={{ caddy_features }} dest={{ caddy_home }}/caddy.tar.gz force=yes
-  when: caddy_releases_cache.changed or caddy_features_cache.changed
-  register: caddy_binary_cache
+  when: caddy_update and (caddy_releases_cache.changed or caddy_features_cache.changed)
```
Comment (lbischof) [inline tasks/main.yml]: Why did you add this check? `caddy_releases_cache` and `caddy_features_cache` already depend on `caddy_update`

File: tasks/main.yml [hunk 2]
```diff
@@ -16,15 +20,21 @@
...
+  register: caddy_downloaded_binary_cache
 
 - name: Download Caddy
   get_url: url=https://caddyserver.com/download/build?os=linux&arch=amd64&features={{ caddy_features }} dest={{ caddy_home }}/caddy.tar.gz
+  when: (not caddy_update) and (caddy_archive_file is not defined)
```
Comment (lbischof) [inline tasks/main.yml]: just for clarification: wouldn't it be enough to check `caddy_archive_file is not defined` because if `caddy_update` is true  or the binary was already downloaded it would just skip this task?

**Comment by lbischof** [comment]
Wouldn't the archive file need to manually be downloaded for this to work?

**Comment by lucafavatella** [comment]
> Wouldn't the archive file need to manually be downloaded for this to work?

Yes. This allows the user invoking the Ansible script to be confident that the invocation is idempotent, as he/she knows which caddy archive is deploying.

E.g. weeks ago I was deploying the archive retrieved at `https://caddyserver.com/download/build?os=linux&arch=amd64&arm=&features=jwt` that at a certain point was returning location `/download/builds/172103150106950/caddy_linux_amd64_custom.tar.gz`.

File: tasks/main.yml (hunk 1 above)
Comment (lucafavatella) [inline tasks/main.yml]: I understand this line of 42364a9faf8152260413d3d2a6aea49fdd9bcc20 is what the commit message part "Also clarify conditions for download." of 42364a9faf8152260413d3d2a6aea49fdd9bcc20 refers to.

I understand that when preparing c45379b627dbad24e1c26efa06a91f43a5434842 I felt that explicating implicit dependencies among variables would have eased readability hence maintenance.

File: tasks/main.yml (hunk 2 above)
Comment (lucafavatella) [inline tasks/main.yml]: Thanks to the assertion at line 8, you are correct. Again, I preferred being explicit as I understand this helps readability and maintainability.

**Comment by lbischof** [comment]
Could you please simplify the conditions (don't check stuff that already relies on another task that does it) and update the readme?

**Comment by lucafavatella** [comment]
Ack. I plan to address this retesting in 15 days or so (busy with other
stuff right now). Thanks for reviewing.

On 26 Apr 2017 10:23, "Lorenz Bischof" <notifications@github.com> wrote:

Could you please simplify the conditions (don't check stuff that already
relies on another task that does it) and update the readme?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/antoiner77/caddy-ansible/pull/42#issuecomment-297315196>,
or mute the thread

**Comment by lucafavatella** [comment]
I plan not to update this anytime soon so closing. Thanks for your review.
